protocol/src/protocol.semanticHighlighting.proposed.md

```diff
@@ -0,0 +1,134 @@
+#### Semantic Highlighting
+
+While the syntax highlighting is done on the client-side and can handle keywords, strings, and other low-level tokens from the grammar, it cannot adequately support complex coloring. Semantic highlighting information is calculated on the language server and pushed to the client as a notification. This notification carries information about the ranges that have to be colored. The desired coloring details are given as [TextMate scopes](https://manual.macromates.com/en/language_grammars) for each affected range. For the semantic highlighting information the following additions are proposed:
+
+_Client Capabilities_:
+
+Capability that has to be set by the language client if it can accept and process the semantic highlighting information received from the server.
+
+```ts
+/**
+ * The text document client capabilities.
+ */
+textDocument?: {
+
+	/**
+	 * The client's semantic highlighting capability.
+	 */
+	semanticHighlightingCapabilities?: {
+
+		/**
+		 * `true` if the client supports semantic highlighting support text documents. Otherwise, `false`. It is `false` by default.
+		 */
+		semanticHighlighting: boolean;
+
+	}
+
+}
+```
+
+_Server Capabilities_:
+
+If the client declares its capabilities with respect to the semantic highlighting feature, and if the server supports this feature too, the server should set all the available TextMate scopes as a "lookup table" during the `initialize` request.
+
+```ts
+/**
+ * Semantic highlighting server capabilities.
+ */
+semanticHighlighting?: {
```

Comment:
Also note that instead of directly referencing a `{ }` instance, you should probably create and refer to a separate, new `SemanticHighlightingOptions` interface kind of like `DocumentLinkOptions` or `DocumentOnTypeFormattingOptions`.